packages/core/src/connectors/metaMask.ts

```diff
@@ -28,6 +29,18 @@ export class MetaMaskConnector extends InjectedConnector {
         ...options,
       },
     })
+
+    // We need this as MetaMask can emit the "disconnect" event
+    // upon switching chains. This workaround ensures that the
+    // user currently isn't in the process of switching chains.
+    const onDisconnect = this.onDisconnect
+    super.onDisconnect = () => {
+      if (this.#switchingChains) {
+        this.#switchingChains = false
+        return
+      }
+      onDisconnect()
+    }
```

Comment:
Think we should expose this behind a `shim` flag? This a more robust version of something I did before in https://github.com/tmm/wagmi/pull/190 (removed since I thought MM fixed the issue), but there could be unintended side effects. Maybe we default the flag to `true` since it seems like something most will want.

Let's link to this issue too: https://github.com/MetaMask/metamask-extension/issues/13375#issuecomment-1087023057

Reply:
Sounds good!